tuf_conformance/test_keys.py

```diff
@@ -8,69 +8,90 @@
-    assert client.version(Root.type) == 4
+    assert client.version(Snapshot.type) == 2
+    assert client.version(Root.type) == 2
 
 
 def test_root_has_keys_but_not_snapshot(
     client: ClientRunner, server: SimulatorServer
 ) -> None:
-    """This test adds keys to the repo root MD to test for cases
-    where are client might calculate the threshold from only the
-    roots keys and not check that the snapshot MD has the same
-    keys"""
+    """This test adds keys to the repo root MD to test for a case
+    where a client might calculate the threshold from only the
+    roots keys and not check that the snapshot MD also has the
+    same keys. For example, the goal is to bring the repository
+    into a state where:
+
+    1. repo.root.roles.snapshot.keyids has the following keyids:
+      1.a: ccce1a69b4eea9daf315d8a9c43fffd8ee541bfb41fdcb773657192af51d3cfa
+      1.b: 4f7c454c55a81918fa1fc1f513d510f36c3efa796711d9f5883602f91a5b9da7
+      1.c: 146e0dc5038d3b8772f5d4bd6626fe1436f656494b501628a2b54d6aa6a6edfe
+      1.d: 3c80449ebd33c67ff8d9befce23c534b50d97ed805e52da8c746e6a1daefab8e
+      1.e: 11ba9251933057a171bac4c1e6d3d9f32ba49fce5dade28a54d5c779538e802d
+    2: repo.root.roles.snapshot.threshold is 5
+    3: repo.snapshot.signatures has the following keyids:
+      3.a: ccce1a69b4eea9daf315d8a9c43fffd8ee541bfb41fdcb773657192af51d3cfa
+      3.b: 4f7c454c55a81918fa1fc1f513d510f36c3efa796711d9f5883602f91a5b9da7
+      3.c: 146e0dc5038d3b8772f5d4bd6626fe1436f656494b501628a2b54d6aa6a6edfe
+      3.d: 3c80449ebd33c67ff8d9befce23c534b50d97ed805e52da8c746e6a1daefab8e
+
+    The test ensures that clients do not update snapshot in this scenario.
+    """
     init_data, repo = server.new_test(client.test_name)
 
     assert client.init_client(init_data) == 0
     assert client.refresh(init_data) == 0
 
     initial_setup_for_key_threshold(client, repo, init_data)
 
-    # Increase the threshold
+    # Increase the threshold to a number higher
+    # than the keys available
     repo.root.roles[Snapshot.type].threshold = 5
-    repo.bump_root_by_one()  # v5
+    repo.bump_root_by_one()  # v3
 
-    # Refresh should fail: root updates but there's no new snapshot:
-    # The existing snapshot does not meet threshold anymore.
+    # Refresh should fail: There is a new root version but
+    # no new snapshot metadata.
+    # The existing snapshot does not meet the threshold
+    # anymore, because there are three keys,
+    # and the threshold is 5.
     # NOTE: we don't actually expect clients to delete the
     # file from trusted_roles() at this point
     assert client.refresh(init_data) == 1
-    assert client.version(Root.type) == 5
+    assert client.version(Root.type) == 3
 
-    # Add two keyids only to root and expect the client
-    # to fail updating
+    # Add one keyid to root but not to snapshot.
     signer = repo.new_signer()
     repo.root.roles[Snapshot.type].keyids.append(signer.public_key.keyid)
```

Comment:
> If you want there to not be 5 signatures in snapshot after this, you can then remove one of the signers from repo.signers

If we remove a signer from `repo.signers`, the repo metadata remains the same. For example, these assertions pass:

```python
repo.root.roles[Snapshot.type].threshold = 4
assert len(repo.root.roles[Snapshot.type].keyids) == 4
assert len(repo.signers[Snapshot.type]) == 4
repo.signers[Snapshot.type].popitem()
assert len(repo.root.roles[Snapshot.type].keyids) == 4
assert len(repo.signers[Snapshot.type]) == 3
repo.bump_root_by_one()  # v3
repo.update_snapshot()  # v3
assert client.refresh(init_data) == 1
```

i.e., the keyids remain the same. 

It works as intended, but it may be confusing when you look at the metadata first.

Reply:
```
repo.add_key()
repo.signers[Snapshot.type].popitem()
```
this should result in 
* root metadata with a snapshot role that has 5 keyids
* snapshot metadata that is signed by 4 keyids

I thought this was what you wanted? As mentioned in the other comment, a new root must be published before client sees it.

Reply:
I like the suggestion of removing instead of adding at that point, since we can significantly reduce the test.

Reply:
> I thought this was what you wanted?

Yes, it is.

> ```
> repo.add_key(Snapshot.type)
> repo.bump_root_by_one()
> repo.signers[Snapshot.type].popitem()
> repo.update_snapshot()
> ```
> this should result in
>
> - root metadata with a snapshot role that has 5 keyids
> - snapshot metadata that is signed by 4 keyids

As far as I can tell, it doesn't. When we remove a signer, we do not remove a signature from `repo.mds[Snapshot.type].signatures`. These 4 assertions all pass:

```python
assert len(repo.signers[Snapshot.type]) == 5
assert len(repo.mds[Snapshot.type].signatures) == 4
repo.signers[Snapshot.type].popitem()
repo.update_snapshot()  # v3
assert len(repo.mds[Snapshot.type].signatures) == 4 # This remains unchanged.
assert len(repo.signers[Snapshot.type]) == 4
```

Reply:
> sure but since the signing still happens at request time, that metadata has not been signed yet here. The signing at request time can make things more confusing than they need to be... but even without it, I think looking at the repository dump in `/tmp/tuf-conformance-dump` may be a better way to verify what client is seeing (than debug assertions about the simulator internals)

Fair enough. It may be my own personal opinion, but I think I in this case would prefer to deduce from the test why it fails.

Reply:
Fixed by https://github.com/theupdateframework/tuf-conformance/pull/160/commits/d064532620148ba02f7f0791bf6db5f3e3b67317